File: web/src/products/curation/research/domain/exchangeRate.test.ts
import { expect, it } from "vitest";
import { convertResearchMinor } from "./exchangeRate";
import { formatMinor } from "../infra/liveCatalogReviewApi";
const now = Date.parse("2026-09-11T03:00:00Z");
const rate = { base: "USD", quote: "KRW", rate: "1340.18", asOf: "2026-09-10", observedAt: "2026-09-11T00:00:00Z", source: "https://frankfurter.dev/" } as const;
it("KRW has zero decimal places, USD has two, and conversion preserves original money", () => {
 const original = { amount: 1234, currency: "USD" };
 expect(convertResearchMinor(original.amount, original.currency, "KRW", rate, now)).toBe(16538);
 expect(convertResearchMinor(23400, "KRW", "USD", rate, now)).toBe(1746);
 expect(original).toEqual({ amount: 1234, currency: "USD" });
 expect(formatMinor(23400, "KRW")).toContain("23,400"); expect(formatMinor(1234, "USD")).toContain("12.34");
});
it("missing, future, stale and unsafe values never become displayed estimates", () => {
 for (const patch of [{ asOf: "2026-09-01" }, { asOf: "2026-09-12" }, { rate: "0" }, { rate: "NaN" }]) expect(convertResearchMinor(100, "USD", "KRW", { ...rate, ...patch }, now)).toBeUndefined();
 expect(convertResearchMinor(Number.MAX_SAFE_INTEGER + 1, "USD", "KRW", rate, now)).toBeUndefined();
 expect(convertResearchMinor(1.2, "KRW", "USD", rate, now)).toBeUndefined();
});

File: server/internal/curation/research/domain/exchange_rate_test.go
package domain

import (
	"testing"
	"time"
)

func TestResearchExchangeConversionRespectsMinorUnitsAndStaleness(t *testing.T) {
	now := time.Date(2026, 9, 11, 12, 0, 0, 0, time.UTC)
	rate := DailyExchangeRate{Base: "USD", Quote: "KRW", Rate: "1340.18", AsOf: "2026-09-10", ObservedAt: now}
	for _, row := range []struct {
		amount   int64
		from, to string
		want     int64
	}{{100, "USD", "KRW", 1340}, {1234, "USD", "KRW", 16538}, {134018, "KRW", "USD", 10000}, {999, "KRW", "KRW", 999}} {
		got, err := ConvertResearchMinor(row.amount, row.from, row.to, rate, now)
		if err != nil || got != row.want {
			t.Fatalf("%+v got=%d err=%v", row, got, err)
		}
	}
	rate.AsOf = "2026-09-01"
	if _, err := ConvertResearchMinor(100, "USD", "KRW", rate, now); err == nil {
		t.Fatal("expired rate accepted")
	}
	rate.AsOf = "2026-09-10"
	rate.Rate = "0"
	if _, err := ConvertResearchMinor(100, "KRW", "USD", rate, now); err == nil {
		t.Fatal("zero divisor accepted")
	}
}
